Append details on parameters to expanded function docstrings.

smspillaz opened the pull request:
Test runners will occassionaly use the docstring of a test method
as a means of reporting its execution. Because the docstring is
simply copied from the template method to the parameterized
method, it isn't very useful in distinguishing between tests.

We now detect the names of the parameters of the function and
the arguments that are being passed as parameters and append
that information to the end of the first line of the docstring.

The generated docstrings might not make perfect sense, so the
user can also specify testcase_func_doc to override docstring
generation with information that makes more sense.

Fixes #21

smspillaz commented:
Bump?

wolever commented:
D'oh! I'm very sorry! I thought I'd commented on this already. Anyway, I like the idea! See comments on the PR.

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -228,7 +270,15 @@ def parameterized_expand_wrapper(f):
                         name_suffix += "_" + cls.to_safe_name(p.args[0])
                     name = base_name + name_suffix
+
+                if testcase_func_doc:
```
Please refactor this so that what is now the `doc_expansion` function is defined at the module level, called something like `default_testcase_func_doc`, and has the same signature as the user-supplied testcase doc func would have (ie, so this if-statement is removed and there's just one line which is `testcase_func_doc = testcase_func_doc or default_testcase_func_doc`.

smspillaz replied:
Yup, sounds good.

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
+            #
+            # We can assume that we are not using parameterized.expand
+            # on a static method.
+            is_method = frame_locals.get('__module__') is not None
```
Is it even possible to use `@parameterized.expand` outside of test classes? IMO it'd be acceptable to skip this check and simply use `args = inspect.getargspec(f).args; if args[0] == "self": args = args[1:]`.

smspillaz replied:
Perhaps it'd be a good idea to assert/throw an exception/add a warning if we're not using this inside a method then? Just in case someone tries to use `expand` on a free function?

The only other thing I can think of is the corner case where you are bound to a method and the first argument is not `self`, but something else. I've seen some ancient code that does that.

If we only want to support bound methods, then we should probably sanity check that the first argument is `self` and if it isn't then complain, preferably loudly.

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
-        def parameterized_expand_wrapper(f):
+        def doc_expansion(f, args, kwargs, is_method):
+            offset = 1 if is_method else 0
+            vns = inspect.getargspec(f)[0][offset:]
```
We can assume 2.6+ so we can use the namedtuple that `getargspec(…)` returns: `argspec = inspect.getargspec(f); args = argspec.args`

smspillaz replied:
Okay.

I'm not sure why I added this here - I suspect I was running into an exception locally, although I'm using 2.7 and 3.4 here. I'll have to dig into it and let you know.

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
+            descs.extend(["{0} = {1}".format(k, repr(v)) for k, v in kwargs.items()])
+
+            # Make up for the shortfall by setting the rest to None
+            descs.extend(["{0} = {1}".format(n, "None") for n in vns[len(args) + len(kwargs.keys()):]])
```
This is incorrect when the default value is not `None` (ex, `def foo(self, bar=42)`).

smspillaz replied:
Hmm. Is there a way to look up the default value in an argspec?

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
+            offset = 1 if is_method else 0
+            vns = inspect.getargspec(f)[0][offset:]
+            descs = ["{0} = {1}".format(n, repr(v)) for n, v in zip(vns, args)]
+            descs.extend(["{0} = {1}".format(k, repr(v)) for k, v in kwargs.items()])
```
Please first merge the provided kwargs with the default kwargs then sort those, so the output order will be consistent both between testcases and between test runs (ex, so `def test_foo(self, a=1, b=2)` run with `()`, `a=42`, and `b=42`, will have `a = 1, b = 2`, `a = 42, b = 2`, and `a = 1, b = 42` respectively).

smspillaz replied:
+1

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
+        def doc_expansion(f, args, kwargs, is_method):
+            offset = 1 if is_method else 0
+            vns = inspect.getargspec(f)[0][offset:]
+            descs = ["{0} = {1}".format(n, repr(v)) for n, v in zip(vns, args)]
```
Instead of using straight repr here, can we get a bit fancy and add a `short_repr` function, which would basically be:

```
def short_repr(x, ln=64):
    x_repr = repr(x)
    if len(x_repr) > n:
        x_repr = x_repr[:n/2] + "..." + x_repr[len(x_repr) - n/2:]
    return x_repr

wolever replied:
This is also going to get crashy with mixed str / unicode (I don't think it's unreasonable to assume that there might be unicode in a docstring, and we need to assume that reprs will return some asinine combination of bytes and unicode).

I believe the fix will be:

1. Tweaking `short_repr` to guarantee that it always returns `unicode` (see below), and
2. Using %-formatting or `u` prefixes on the `.format` strings (https://gist.github.com/wolever/be5f554d06d666471283).

```
def short_repr(x, ln=64):
    x_repr = repr(x)
    if isinstance(x_repr, str):
        try:
            x_repr = unicode(x_repr, "utf-8")
        except UnicodeDecodeError:
            x_repr = unicode(x_repr, "latin1")
    if len(x_repr) > n:
        x_repr = x_repr[:n/2] + "..." + x_repr[len(x_repr) - n/2:]
    return x_repr
```

smspillaz replied:
Noted.

I'll add some mixed-unicode testcases too. Unicode is just annoyingly painful to get right a lot of the time :(

wolever commented on nose_parameterized/parameterized.py:
```diff
@@ -211,11 +211,53 @@ def expand(cls, input, testcase_func_name=None):
+                first = first + " [with {0}]".format(", ".join(descs)) + append
+                return "\n".join([first] + split_doc[1:])
+            else:
+                return None
```
Why not assume an empty doc if `f.__doc__` is `None`?

smspillaz replied:
Hmmm, I'm not sure what the best thing to do here is.

I know that test runners which do look at the function's `__doc__`  will default to showing the function name instead of the documentation if `__doc__` is set to None. I suspect those same test runners don't treat `None` and `""` in the same way, which could lead to blank-line-tests.

smspillaz commented:
@wolever  No problem, sometimes things fall of radars :)

I'll respond to the comments with any relevant general ideas, although I won't be able to hack on this for a few days. Hope that's okay.

At the moment I'm using a personal fork of this for my own projects - (here)[https://travis-ci.org/polysquare/polysquare-cmake-linter/jobs/66171425] is what I had in mind, in case you were curious.

smspillaz commented on setup.py:
```diff
@@ -9,7 +9,7 @@
 setup(
     name="nose-parameterized",
-    version="0.3.5",
+    version="0.4.0",
```
Oops. I think this version bump was an unintentional artifact.

Do you want me to leave this here or get rid of it for now?

wolever commented:
Woo! Okay, looking good.

I've made a few tweaks to the argument pairs function (it turns out to be nontrivial to correctly line up and order args, kwargs, defaults, *varargs, and **keywords) in https://github.com/wolever/nose-parameterized/tree/pull-22

The last thing before this gets merged: the test cases you've added are very verbose and it's not clear what they are testing. Please simplify them so there's no duplication and it's obvious what's being tested in each one. I'd imagine doing something like this:

```python

class TestParameterizedExpandDocstring(object):
         @parameterized.expand([("foo", )])
         def test_no_docstring(self, arg):
             assert_testcase_docstring(None)

         @parameterized.expand([("foo", )])
         def test_unicode_docstring(self, arg):
             u"ƒøö"
             assert_testcase_docstring(u'ƒøö [with arg="foo"]')
```

And this will be good to merge.

Oh, and keep the version number — I think this deserves a minor version bump.

smspillaz commented:
@wolever 

Yeah I was going to mention that the current "schema" for test cases is a little vague.

Do you mind if I add more detailed names to all of those test cases in that class?

smspillaz commented:
Oh, I was going to ask - should we do anything about `parameterized.expand` being called on a non-bound method? I noticed that  919a4931a3f2dcbec1ebb6f7460cd82f555ab388 removed the check.

smspillaz commented:
Okay, cleaned up test cases as suggested.

I noticed quite a few tests were failing too, so I've fixed what was causing that to happen.

I'm still not sure what the best thing to do about instance methods vs free functions. I've just added a check for "self" being the first argument and adjusting silently from there, but it will break if an instance
method uses something other than self for the first argument.

wolever commented:
Oh crap, yes, I forgot about the "self" arg offset. Thanks!

Okay! Looking good. I've fixed the Py3 failures and will be pushing a new release soon.

Thanks!

smspillaz commented:
@wolever No problem. Thanks for fixing the py3 failures too - I'm not sure how to get tox to pick up on the fact that I've got a py3 installation on OS X, hopefully one day I'll figure that out.
